platform/CircularBuffer.h

```diff
@@ -30,21 +30,25 @@ namespace mbed {
  *
  *  @note Synchronization level: Interrupt safe
  */
-template<typename T, uint32_t BufferSize, typename CounterType = uint32_t>
+template<typename T, uint32_t BufferSize>
```

Comment:
That's a breaking change; why not use a static assertion rather than removing the template parameter: 

```C++
    MBED_STATIC_ASSERT(
        (sizeof(CounterType) >= sizeof(uint32_t)) ||
        (BufferSize < (((uint64_t) 1) << (sizeof(CounterType) * 8))),
        "Invalid BufferSize for the CounterType"
    );
```

Reply:
Assertion is an alternative to this solution. I decided to make a proposal with removed template parameter since I can't see any advantage to have it. Why do we need it?

Provided solution is not complete. Also signed counter types should be forbidden. The following declaration is also invalid and will not be caught by the static assertion:
`CircularBuffer<int, 255, char> cb;`

Reply:
`CounterType` allows user of the class to choose the most appropriate variable type for inner counters of the object; it might save few bytes by `CircularBuffer`. It would have been better if the type selection was made automatically based on the buffer size but it is here now and we have to live with it.  

> Provided solution is not complete. Also signed counter types should be forbidden. The following declaration is also invalid and will not be caught by the static assertion:
CircularBuffer<int, 255, char> cb;

The static assertion can handle the sign of the `CounterType` if you define a type traits that determine if an integral type is signed or undisgned.

Reply:
We are committed to preserving backward compatibility for our user facing APIs for 5.x versions of Mbed OS.